File: app/src/main/java/me/instabattle/app/models/Entry.java
package me.instabattle.app.models;

import android.graphics.Bitmap;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.util.Date;

import me.instabattle.app.managers.BattleManager;
import me.instabattle.app.managers.BitmapCallback;
import me.instabattle.app.managers.PhotoManager;
import me.instabattle.app.managers.UserManager;
import retrofit2.Callback;

public class Entry {
    private static int RATING_MULTIPLIER = 60;

    @SerializedName("id")
    @Expose
    private Integer id;
    @SerializedName("user_id")
    @Expose
    private Integer authorId;
    @SerializedName("battle_id")
    @Expose
    private Integer battleId;
    @SerializedName("rating")
    @Expose
    private Double rating = 25d;
    @SerializedName("created_on")
    @Expose
    private Date createdOn;
    @SerializedName("image")
    @Expose
    private String imageName;
    @SerializedName("latitude")
    @Expose
    private Double lat;
    @SerializedName("longitude")
    @Expose
    private Double lng;
    //FIXME
    private String description;


    public Entry(int battleId, int authorId, double lat, double lng) {
        this.authorId = authorId;
        this.battleId = battleId;
        this.lat = lat;
        this.lng = lng;
    }

    public void getPhotoAndDo(BitmapCallback callback) {
        PhotoManager.getPhotoAndDo(this.getImageName(), callback);
    }

    public void getAuthorAndDo(Callback<User> callback) {
        //UserManager.getAndDo(authorId, callback);
    }

    public int getRating() {
        return (int) (rating * RATING_MULTIPLIER);
    }

    public Integer getId() {
        return id;
    }

    public String getImageName() {
        return imageName;
    }

    public void getBattleAndDo(Callback<Battle> callback) {
        BattleManager.getAndDo(battleId, callback);
    }

    public void setId(int id) {
        this.id = id;
    }

    public Date getCreatedOn() {
        return createdOn;
    }

    public Integer getBattleId() {
        return battleId;
    }

    public Integer getAuhtorId() {
        return authorId;
    }
}

File: app/src/main/kotlin/me/instabattle/app/models/Entry.kt
package me.instabattle.app.models

import android.os.Parcelable
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import kotlinx.android.parcel.Parcelize
import me.instabattle.app.managers.BattleManager
import me.instabattle.app.managers.UserManager
import retrofit2.Callback
import java.util.*

@Parcelize
data class Entry(
        @field:SerializedName("battle_id") @field:Expose
        val battleId: Int,
        @field:SerializedName("user_id") @field:Expose
        val authorId: Int?,
        @field:SerializedName("latitude") @field:Expose
        private val lat: Double,
        @field:SerializedName("longitude") @field:Expose
        private val lng: Double,
        @field:SerializedName("id") @field:Expose
        var id: Int? = null,
        @field:SerializedName("rating") @field:Expose
        val rating: Double? = 25.0,
        @field:SerializedName("created_on") @field:Expose
        val createdOn: Date? = null,
        @field:SerializedName("image") @field:Expose
        val imageName: String? = null,
        @field:SerializedName("description") @field:Expose
        val description: String? = null) : Parcelable {

    fun getAuthorAndDo(callback: Callback<User>) {
        UserManager.getAndDo(authorId, callback)
    }

    fun getRating(): Int {
        return (rating!! * RATING_MULTIPLIER).toInt()
    }

    fun getBattleAndDo(callback: Callback<Battle>) {
        BattleManager.getAndDo(battleId, callback)
    }

    companion object {
        private const val RATING_MULTIPLIER = 60
    }
}
